pkg/utils/keypath/key_path_v2.go

```diff
@@ -0,0 +1,101 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keypath
+
+import (
+	"fmt"
+	"path"
+)
+
+const (
+	leaderPathFormat              = "/pd/%d/leader"                   // "/pd/{cluster_id}/leader"
+	dcLocationPathFormat          = "/pd/%d/dc-location/%d"           // "/pd/{cluster_id}/dc-location/{member_id}"
+	memberBinaryDeployPathFormat  = "/pd/%d/member/%d/deploy_path"    // "/pd/{cluster_id}/member/{member_id}/deploy_path"
+	memberGitHashPath             = "/pd/%d/member/%d/git_hash"       // "/pd/{cluster_id}/member/{member_id}/git_hash"
+	memberBinaryVersionPathFormat = "/pd/%d/member/%d/binary_version" // "/pd/{cluster_id}/member/{member_id}/binary_version"
+	allocIDPathFormat             = "/pd/%d/alloc_id"                 // "/pd/{cluster_id}/alloc_id"
+	keyspaceAllocIDPathFormat     = "/pd/%d/keyspaces/alloc_id"       // "/pd/{cluster_id}/keyspaces/alloc_id"
+
+	msLeaderPathFormat           = "/ms/%d/%s/primary"                                       // "/ms/{cluster_id}/{service_name}/primary"
+	msTsoDefaultLeaderPathFormat = "/ms/%d/tso/00000/primary"                                // "/ms/{cluster_id}/tso/00000/primary"
+	msTsoKespaceLeaderPathFormat = "/ms/%d/tso/keyspace_groups/election/%05d/primary"        // "/ms/{cluster_id}/tso/keyspace_groups/election/{group_id}/primary"
+	msDCLocationPathFormat       = "/ms/%d/%s/dc-location/%d"                                // "/ms/{cluster_id}/{service_name}/dc-location/{member_id}"
+	msTsoDefaultDCLocationPath   = "/ms/%d/tso/00000/dc-location/%d"                         // "/ms/{cluster_id}/tso/00000/dc-location/{member_id}"
+	msTsoKespaceDCLocationPath   = "/ms/%d/tso/keyspace_groups/election/%05d/dc-location/%d" // "/ms/{cluster_id}/tso/keyspace_groups/election/{group_id}/dc-location/{member_id}"
+)
+
+// MsParam is the parameter of micro service.
+type MsParam struct {
+	ServiceName string
+	GroupID     uint32 // only used for tso keyspace group
+}
+
+// Prefix returns the parent directory of the given path.
+func Prefix(str string) string {
+	return path.Dir(str)
+}
+
+// LeaderPath returns the leader path.
+func LeaderPath(p *MsParam) string {
+	if p == nil || p.ServiceName == "" {
+		return fmt.Sprintf(leaderPathFormat, ClusterID())
+	}
+	if p.ServiceName == "tso" {
+		if p.GroupID == 0 {
+			return fmt.Sprintf(msTsoDefaultLeaderPathFormat, ClusterID())
+		}
+		return fmt.Sprintf(msTsoKespaceLeaderPathFormat, ClusterID(), p.GroupID)
+	}
+	return fmt.Sprintf(msLeaderPathFormat, ClusterID(), p.ServiceName)
+}
+
+// DCLocationPath returns the dc-location path.
+func DCLocationPath(p *MsParam, memberID uint64) string {
```

Comment:
Can we just remove them?

Reply:
Considering the server side code of Local TSO is not removed yet, it is better to keep them now, and remove them later